File: src/aws/lambdas/incremental_distribution/src/main/java/uk/nhs/nhsx/core/aws/s3/AwsS3Client.java
package uk.nhs.nhsx.core.aws.s3;

import com.amazonaws.services.s3.AmazonS3;
import com.amazonaws.services.s3.AmazonS3ClientBuilder;
import com.amazonaws.services.s3.model.*;
import com.google.common.io.ByteSource;
import org.apache.http.entity.ContentType;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.Optional;

import static java.util.Arrays.stream;

public class AwsS3Client implements AwsS3 {

    private static final Logger logger = LogManager.getLogger(AwsS3Client.class);

    private final AmazonS3 client;

    public AwsS3Client() {
        client = AmazonS3ClientBuilder.defaultClient();
    }

    @Override
    @SuppressWarnings("UnstableApiUsage")
    public void upload(Locator locator, ContentType contentType, ByteSource bytes, MetaHeader... meta) {
        ObjectMetadata metadata = new ObjectMetadata();
        metadata.setContentType(contentType.getMimeType());
        stream(meta).forEach(m -> metadata.addUserMetadata(m.asS3MetaName(), m.value));
        bytes.sizeIfKnown().toJavaUtil().ifPresent(metadata::setContentLength);
        try (InputStream input = bytes.openBufferedStream()) {
            client.putObject(
                new PutObjectRequest(locator.bucket.value, locator.key.value, input, metadata)
            );
        } catch (IOException e) {
            throw new RuntimeException(String.format("Unable to open data source %s", bytes), e);
        }
    }

    @Override
    public List<S3ObjectSummary> getObjectSummaries(BucketName bucketName) {
        ListObjectsV2Request listObjectsV2Request = new ListObjectsV2Request();
        listObjectsV2Request.setBucketName(bucketName.value);

        ListObjectsV2Result result = client.listObjectsV2(listObjectsV2Request);
        List<S3ObjectSummary> objectSummaries = new ArrayList<>(result.getObjectSummaries());

        while (result.isTruncated()) {
            String token = result.getNextContinuationToken();
            listObjectsV2Request.setContinuationToken(token);

            result = client.listObjectsV2(listObjectsV2Request);
            objectSummaries.addAll(result.getObjectSummaries());
        }

        return objectSummaries;
    }

    @Override
    public Optional<S3Object> getObject(Locator locator) {
        try {
            return Optional.ofNullable(client.getObject(locator.bucket.value, locator.key.value));
        } catch (AmazonS3Exception e) {
            if (!(e.getStatusCode() == 404 && Objects.equals(e.getErrorCode(), "NoSuchKey"))) {
                logger.warn("Object could not be retrieved", e);
            }
            return Optional.empty();
        }
    }

    @Override
    public void deleteObject(Locator locator) {
        client.deleteObject(locator.bucket.value, locator.key.value);
    }

}

File: src/aws/lambdas/incremental_distribution/src/main/java/uk/nhs/nhsx/core/aws/s3/AwsS3Client.kt
package uk.nhs.nhsx.core.aws.s3

import com.amazonaws.services.s3.AmazonS3
import com.amazonaws.services.s3.AmazonS3ClientBuilder
import com.amazonaws.services.s3.model.AmazonS3Exception
import com.amazonaws.services.s3.model.ListObjectsV2Request
import com.amazonaws.services.s3.model.ObjectMetadata
import com.amazonaws.services.s3.model.PutObjectRequest
import com.amazonaws.services.s3.model.S3Object
import com.amazonaws.services.s3.model.S3ObjectSummary
import org.apache.http.entity.ContentType
import uk.nhs.nhsx.core.events.Events
import java.io.IOException
import java.net.URL
import java.util.*

class AwsS3Client(private val events: Events) : AwsS3 {

    private val client: AmazonS3 = AmazonS3ClientBuilder.defaultClient()

    override fun upload(
        locator: Locator,
        contentType: ContentType,
        bytes: ByteArraySource,
        meta: List<MetaHeader>
    ) {
        events.emit(javaClass, S3Upload(locator.key, locator.bucket))

        val metadata = ObjectMetadata().apply {
            this.contentType = contentType.mimeType
            this.contentLength = bytes.size.toLong()
            meta.forEach { addUserMetadata(it.asS3MetaName(), it.value) }
        }

        try {
            bytes.openStream().use {
                client.putObject(PutObjectRequest(locator.bucket.value, locator.key.value, it, metadata))
            }
        } catch (e: IOException) {
            throw RuntimeException("Unable to open data source", e)
        }
    }

    override fun getObjectSummaries(bucketName: BucketName): List<S3ObjectSummary> {
        val request = ListObjectsV2Request().apply {
            this.bucketName = bucketName.value
        }

        var result = client.listObjectsV2(request)

        val objectSummaries = result.objectSummaries?.toMutableList() ?: mutableListOf()

        while (result.isTruncated) {
            request.continuationToken = result.nextContinuationToken
            result = client.listObjectsV2(request)
            objectSummaries.addAll(result.objectSummaries)
        }

        return objectSummaries
    }

    override fun getObject(locator: Locator): Optional<S3Object> {
        return try {
            Optional.ofNullable(client.getObject(locator.bucket.value, locator.key.value))
        } catch (e: AmazonS3Exception) {
            if (!e.is404()) {
                events.emit(javaClass, S3Error(locator, e.statusCode, e.errorCode))
            }
            Optional.empty()
        }
    }



    private fun AmazonS3Exception.is404() = statusCode == 404 && errorCode == "NoSuchKey"

    override fun deleteObject(locator: Locator) = client.deleteObject(locator.bucket.value, locator.key.value)

    override fun getSignedURL(locator: Locator, expiration: Date): Optional<URL> {
        return try {
            Optional.ofNullable(client.generatePresignedUrl(locator.bucket.value,locator.key.value,expiration))
        } catch (e: AmazonS3Exception) {
            if (!e.is404()) {
                events.emit(javaClass, S3Error(locator, e.statusCode, e.errorCode))
            }
            Optional.empty()
        }
    }
}
